### PR TITLE
[FIX] properly infer slice_time_ref from BIDS derivatives

### LINKED ISSUE #3299: first_level_from_bids warnings ignore slice_time_ref parameter
Hello - 

when running _first_level_from_bids_ structure with specified float valued slice_time_ref (e.g: 0.5 for default fmriprep'd data) the following warning:
    if t_r is not None:
        warn('RepetitionTime given in model_init as %d' % t_r)
        warn('slice_time_ref is %d percent of the repetition '
             'time' % slice_time_ref)
%d rounds the values to zero.

In addition 
> UserWarning: SliceTimingRef not found in file C:\Users\saptaf1/nilearn_data\ds000030/ds000030_R1.0.4/uncompressed\sub-10159\func\sub-10159_task-stopsignal_bold.json. It will be assumed that the slice timing reference is 0.0 percent of the repetition time. If it is not the case it will need to be set manually in the generated list of models
ignores specified slice_time_ref argument.

Nilearn version:
0.9.1
### Expected behavior
1.Display slice_time_ref (and the TR) as float in user warning
2.Don't ignore slice_time_ref argument in warning message display logic 

### Actual behavior
TR and slice_time_ref displayed as decimal integers in first warning and slice_time_ref argument ignored in the second warning.

### Steps and code to reproduce bug

Note this code also triggers the 2nd issue, because the TR is extracted from the json.

```python

from nilearn.datasets import (fetch_openneuro_dataset_index,
                              fetch_openneuro_dataset, select_from_index)

_, urls = fetch_openneuro_dataset_index()

exclusion_patterns = ['*group*', '*phenotype*', '*mriqc*',
                      '*parameter_plots*', '*physio_plots*',
                      '*space-fsaverage*', '*space-T1w*',
                      '*dwi*', '*beh*', '*task-bart*',
                      '*task-rest*', '*task-scap*', '*task-task*']
urls = select_from_index(
    urls, exclusion_filters=exclusion_patterns, n_subjects=1)

data_dir, _ = fetch_openneuro_dataset(urls=urls)

from nilearn.glm.first_level import first_level_from_bids
task_label = 'stopsignal'
space_label = 'MNI152NLin2009cAsym'
derivatives_folder = 'derivatives/fmriprep'
models, models_run_imgs, models_events, models_confounds = \
    first_level_from_bids(data_dir, task_label, space_label,
                          smoothing_fwhm=5.0,
                          derivatives_folder=derivatives_folder, slice_time_ref=0.5)

```

### PR DESCRIPTION
**Pull request by Remi-Gau**
Closes #3299

Changes proposed in this pull request:

- decouple fetching `RepetitionTime` from fetching metadata related to slice timing
- start a deprecating cycle to change the default value of slice_time_ref from 0 to None
- add `t_r` and `slice_time_ref` validation for first level models
- extract functions to get metadata from bids sidecar json files
- add appropriate tests including regression test and test for deprecation warning

### CODE REVIEW

File: nilearn/_utils/data_gen.py
Comment (tsalo) [inline nilearn/_utils/data_gen.py:753]: The spacing looks off here. Any clue why flake8 didn't catch it?

```suggestion
def add_metadata_to_bids_derivatives(bids_path: str | Path,
                                     metadata: dict,
                                     json_file: str = None) -> Path:
```

Comment (tsalo) [reply nilearn/_utils/data_gen.py:753]: Also, shouldn't `black` reformat it to something like the following?

```suggestion
def add_metadata_to_bids_derivatives(
    bids_path: str | Path,
    metadata: dict,
    json_file: str = None,
) -> Path:
```

Comment (Remi-Gau) [reply nilearn/_utils/data_gen.py:753]: that's because flake8 and black are not yet active on those files

trying to keep a list of things that are covered here: https://github.com/nilearn/nilearn/projects/8#card-88537577

Otherwise check the long exception list in here to see what black works or does not work on.

https://github.com/nilearn/nilearn/blob/a744d100c52d6844f643343797eb36d08e04470d/pyproject.toml

so while reformatting is happening, expect some more "liberal" formatting.

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:1002]: remove

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:1167]: I am having to pass this verbose parameter around a bit too much to be able to control whether warnings, messages should be printed...

This should probably be handled better by implementing a logger in another PR.

File: nilearn/_utils/data_gen.py
Comment (bthirion) [inline nilearn/_utils/data_gen.py:755]: I thought we were holding off for now

File: nilearn/glm/first_level/first_level.py
Comment (bthirion) [inline nilearn/glm/first_level/first_level.py:834]: Amigos, we need a deprecation warning before changing the API !

File: nilearn/glm/first_level/first_level.py
Comment (bthirion) [inline nilearn/glm/first_level/first_level.py:930]: I don't think we should have a warning here.

File: nilearn/glm/first_level/first_level.py
Comment (bthirion) [inline nilearn/glm/first_level/first_level.py:961]: Please test the warning.

File: nilearn/glm/first_level/first_level.py
Comment (bthirion) [inline nilearn/glm/first_level/first_level.py:965]: We should not warn here.

File: nilearn/glm/tests/test_first_level.py
Comment (bthirion) [inline nilearn/glm/tests/test_first_level.py:609]:
```suggestion
    with pytest.raises(error_type, match=error_msg):
```

File: nilearn/interfaces/bids/query.py
Comment (bthirion) [inline nilearn/interfaces/bids/query.py:52]: Should be a warning or an error ?

File: nilearn/interfaces/bids/query.py
Comment (bthirion) [inline nilearn/interfaces/bids/query.py:99]: Same question here.

File: nilearn/interfaces/bids/query.py
Comment (bthirion) [inline nilearn/interfaces/bids/query.py:143]: and here.

File: nilearn/interfaces/bids/query.py
Comment (bthirion) [inline nilearn/interfaces/bids/query.py:88]: Shouldn't this exist anyhow ?

**Review by bthirion — COMMENTED** [review]
Thx. Please find a few comments.
The main concern is about API change.

File: nilearn/_utils/data_gen.py
Comment (Remi-Gau) [reply nilearn/_utils/data_gen.py:755]: you are correct
I had added them before we made a decision: I thought I had removed all of them but I missed a fews

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [reply nilearn/glm/first_level/first_level.py:834]: I actually need to maybe rethink whether the change of API is even desirable / necessary.

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [reply nilearn/glm/first_level/first_level.py:965]: Gonna try to make a small table about the different situations (input argument VS what metadata exist in which dataset) to clarify when to throw warnings or errors.

Comment (ymzayek) [reply nilearn/glm/first_level/first_level.py:965]: This refactoring isn't very useful if `slice_time_ref` default is not changed to None. If we do this, we need to deprecate it properly as @bthirion pointed out. In any case, given the explanation in the docstrings, this warning becomes redundant with the one under it.

**Comment by Remi-Gau** [comment]
tagging @ymzayek @bthirion @emdupre @tsalo as you are the ones who have interacted with this PR

I tried to write some sort of "behavior requirements" regarding this PR: should probably have started with this rather than getting coding guns blazing, but better later than never.

This is meant to help us decide when to throw warning / errors and also help decide what the default values should be (and hence if we need to change the API).

Can you have a look at the tables below and let me know if you agree with their content?

## `t_r` related behaviors

### `t_r` can be inferred from BIDS

 inputs | behavior when `t_r` inferred BIDS derivatives | behavior when `t_r` inferred from BIDS raw
-- | -- | --
`t_r = None` (default) | t_r value of derivative takes precedence; no error; no warning | use `t_r` value of raw; no error; no warning
t_r provided matches value found in BIDS | no error; no warning | no error; no warning
t_r provided does not match value found in BIDS | throw an error | throw an error

Note that we may want to let users use the "wrong" repetition time and just throw a warning in case of mismatch between the `RepetitionTime` found in the BIDS dataset and the `t_r` value provided.

### `t_r` cannot be inferred from BIDS raw or derivatives

This should be rare as `RepetitionTime` is a required metadata for BIDS raw datasets. This may happen if somehow the relevant JSON file not picked by nilearn `get_bids_file` (possibly because of the BIDS inheritance principle?).

 inputs | behavior
-- | --
t_r = None | throw warning no t_r found: a t_r will need to be provided
t_r provided | no error, no warning

## `slice_timing_ref` related behaviors

### `slice_timing_ref` can be inferred from BIDS

Should be the case for datasets from fmriprep >= 20.2.4

inputs | behavior |
-- | -- |
slice_time_ref = None | use slice_timing_ref of BIDS derivatives |
slice_time_ref provided; matches value inferred from BIDS | no error, no warning |
slice_time_ref provided; does not match value inferred from BIDS | throw an error |

Note that we may want to let users use the "wrong" slice_time_ref and just throw a warning in case of mismatch between the `slice_time_ref` inferred from the BIDS dataset and the `t_r` value provided.

### `slice_timing_ref` cannot be inferred from BIDS

This may happen with (from most likely to more edge case):
- "older" fmriprep datasets (< 20.2.4)
- with processing pipelines that do not apply the suggestions from this [BIDS issue](https://github.com/bids-standard/bids-specification/issues/836) (as this is not yet official BIDS)
- or if the the `StartTime` is in JSON in the BIDS dataset that is not somehow not picked by nilearn `get_bids_file` (possibly because of the BIDS inheritance principle?).

inputs | behavior |
-- | -- |
slice_time_ref = None | set slice_timing_ref to 0
slice_time_ref provided | no error, no warning

### Change API default value

We could decide to NOT change the default value of `slice_timing_ref` to None but in this case users would have to "opt-in" to let nilearn infer the its value by actively passing `slice_timing_ref=None`.

I suspect many users may not be aware of those BIDS / fmriprep subtleties, so having something like this be opt-in rather than a default feels undesirable from my point of view.

Curious to hear others' thoughts.

## TODOs

- [ ] try to infer t_r and slice_time_ref even if they are provided through the arguments
- [ ] implement error when inferred t_r differs from provided t_r
- [ ] implement error when inferred slice_time_ref differs from provided slice_time_ref
- [ ] test ability of get_bids_file to cope with the BIDS inheritance principle (not directly needed for this PR but useful to know what are the "blind spots" of this function)

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [reply nilearn/glm/first_level/first_level.py:965]: will remove the warning

File: doc/changes/latest.rst
Comment (Remi-Gau) [reply doc/changes/latest.rst:19]: yup totally what happened.
thanks for spotting it

File: nilearn/interfaces/bids/query.py
Comment (Remi-Gau) [reply nilearn/interfaces/bids/query.py:52]: I think I would stick to warning here: see my new comment in the PR.


https://github.com/nilearn/nilearn/pull/3605#issuecomment-1490567765

I think the errors should be thrown by some other part of the code depending on what this function returns and what argument was passed to `first_level_from_bids`.

Throwing may help users understand why a subsequent error happened but missing json may not be a sufficient reason to throw an error.

File: nilearn/_utils/data_gen.py
Comment (Remi-Gau) [reply nilearn/_utils/data_gen.py:1]: Those are much more related to the generation of dummy BIDS datasets rather than being usable on actual BIDS dataset.

Does that make sense?

File: nilearn/interfaces/tests/test_bids.py
Comment (Remi-Gau) [inline nilearn/interfaces/tests/test_bids.py:154]: I have not seen xfail used in the test suite so not sure if there is any guideline about using them. 

https://docs.pytest.org/en/7.1.x/how-to/skipping.html#xfail-mark-test-functions-as-expected-to-fail

Will make a strict xfail so that an XPASS (“unexpectedly passing”) results from this test will fail the test suite.

**Comment by bthirion** [comment]
I think I agree with your proposition: I'm also hesitating on the case of incompatibility between inferred and provided value, for both TR and slice_time_ref. I think I'd rather spit a warning to let the user know he/she's
doing something strange, but I would still allow it.

File: nilearn/glm/first_level/first_level.py
Comment (ymzayek) [reply nilearn/glm/first_level/first_level.py]: @Remi-Gau Yes I agree that if we don't have to change the API, better not. But this implementation will only infer the slice_time_ref if the user passes None, while the old behavior tries to infer it if the default, which is `0.`, is kept. So a user that doesn't change their code that uses the default option will now get `0.` set as their slice_time_ref while they expect it to be inferred. Does that make sense? Or maybe I have the logic wrong. If what I'm saying follows, then we'd need to go through a deprecation cycle as well.

File: nilearn/_utils/data_gen.py
Comment (ymzayek) [reply nilearn/_utils/data_gen.py:1]: Yes!

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [reply nilearn/glm/first_level/first_level.py]: If I am correct the old behavior was:
- try to infer `t_r` AND `slice_time_ref` **if and only if** `t_r = None` is passed (which was the default)
- if a `slice_time_ref` is found then it overrides ANY value passed as argument.

Making more tables to compare old and new behavior depending on datasets and arguments...

**Comment by Remi-Gau** [comment]
## comparing old and new behavior

See spreadsheet

https://docs.google.com/spreadsheets/d/1Zco4l6T-Bc9t1m7WfujHCC5c4EJwwm_GH8gU-ADitA8/edit?usp=sharing

Using some "actual" values for what could be inferred from a dataset to make this a bit less abstract.

### TL;DR: 

For users who just rely on the defaults this should not lead to a change in behavior. 

Well except for the fact that the old code was not looking for the right metadata for slice_time_ref in BIDS derivatives dataset, so it was unlikely to give the right answer.

Speaks in favor of:

- start deprecating API for slice_time_ref to start using
- use only warning when there is a mistmatch between passed value and inferred value

### old behavior

Note I think that passing None for slice_time_ref may lead to errors when setting up the first level model but not 100% sure.

The Third line is the default.

  |   | t_r (3) and slice_time_ref (0.75) inferred | t_r and slice_time_ref cannot be inferred | t_r (3) only can be inferred | slice_time_ref (0.75) only can be inferred
-- | -- | -- | -- | -- | --
t_r=None | slice_time_ref=None | t_r=3; slice_time_ref=0.75 | t_r=None; slice_time_ref=None + missing t_r warning | t_r=3; slice_time_ref=None + downstream error? | t_r=None; slice_time_ref=0.75 + missing t_r warning
t_r=2 | slice_time_ref=None | t_r=2; slice_time_ref=None + downstream error? | t_r=2; slice_time_ref=None + downstream error? | t_r=2; slice_time_ref=None + downstream error? | t_r=2; slice_time_ref=None + downstream error?
t_r=None | slice_time_ref=0.0 | t_r=3; slice_time_ref=0.75 | t_r=None; slice_time_ref=0.0 + missing t_r warning | t_r=3; slice_time_ref=0.0 | t_r=None; slice_time_ref=0.75 + missing t_r warning
t_r=2 | slice_time_ref=0.0 | t_r=2; slice_time_ref=0.0 | t_r=2; slice_time_ref=0.0 | t_r=2; slice_time_ref=0.0 | t_r=2; slice_time_ref=0.0
t_r=None | slice_time_ref=0.5 | t_r=3; slice_time_ref=0.75 | t_r=None; slice_time_ref=0.5 + missing t_r warning | t_r=3; slice_time_ref=0.5 | t_r=None; slice_time_ref=0.75 + missing t_r warning
t_r=2 | slice_time_ref=0.5 | t_r=2; slice_time_ref=0.5 | t_r=2; slice_time_ref=0.5 | t_r=2; slice_time_ref=0.5 | t_r=2; slice_time_ref=0.5

### new behavior

The first line would be the new default after the deprecation cycle.

  |   | t_r (3) and slice_time_ref (0.75) inferred | t_r and slice_time_ref cannot be inferred | t_r (3) only can be inferred | slice_time_ref (0.75) only can be inferred
-- | -- | -- | -- | -- | --
t_r=None | slice_time_ref=None | t_r=3; slice_time_ref=0.75 | t_r=None; slice_time_ref=0.0 + missing t_r warning | t_r=3; slice_time_ref=0.0 | t_r=None; slice_time_ref=0.75 + missing t_r warning
t_r=2 | slice_time_ref=None | t_r=2; slice_time_ref=0.75 + mismatch warning | t_r=2; slice_time_ref=0 | t_r=2; slice_time_ref=0 + mismatch warning | t_r=2; slice_time_ref=0.75
t_r=None | slice_time_ref=0.0 | t_r=3; slice_time_ref=0.0 + mismatch warning | t_r=None; slice_time_ref=0.0 + missing t_r warning | t_r=3; slice_time_ref=0.0 | t_r=None; slice_time_ref=0.0 + missing t_r warning + mismatch warning
t_r=2 | slice_time_ref=0.0 | t_r=2; slice_time_ref=0.0 + mismatch warning | t_r=2; slice_time_ref=0.0 | t_r=2; slice_time_ref=0.0 + mismatch warning | t_r=2; slice_time_ref=0.0 + mismatch warning
t_r=None | slice_time_ref=0.5 | t_r=3; slice_time_ref=0.5 + mismatch warning | t_r=None; slice_time_ref=0.5 + missing t_r warning | t_r=3; slice_time_ref=0.5 | t_r=None; slice_time_ref=0.5 + missing t_r warning + mismatch warning
t_r=2 | slice_time_ref=0.5 | t_r=2; slice_time_ref=0.5 + mismatch warning | t_r=2; slice_time_ref=0.5 | t_r=2; slice_time_ref=0.5 + mismatch warning | t_r=2; slice_time_ref=0.5 + mismatch warning

File: doc/changes/latest.rst
Comment (Remi-Gau) [inline doc/changes/latest.rst:18]: not sure what number to put for the deprecation version here

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:894]: not sure what number to put for the deprecation version here

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:925]: not sure what number to put for the deprecation version here

**Review by Remi-Gau — COMMENTED** [review]
implemented the other changes suggested and a deprecation warning

just not sure how long the deprecation cycles are.

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:832]:
```suggestion
                          slice_time_ref=0.0,
```

**Comment by ymzayek** [comment]
@Remi-Gau thanks a lot for these tables! The warning/errors table seems fine to me.

Regarding the second table, I do see that the old vs new default behavior is the same but what I meant was that if users who use the default parameter values but explicitly define them such as `slice_time_ref=0.` in their code then in this (perhaps edge) case, the behavior would be different. In any case, this is all fine if we go through deprecation.

File: doc/changes/latest.rst
Comment (Remi-Gau) [reply doc/changes/latest.rst:18]: OK that makes sense to me

**Comment by Remi-Gau** [comment]
> then in this (perhaps edge) case, the behavior would be different

yup your comment just prompted me to actually try to map most cases to get a better idea because I was not even clear about it myself.

**Comment by Remi-Gau** [comment]
Todo

- [ ] add test to specifically catch the deprecation warning.

File: nilearn/_utils/data_gen.py
Comment (bthirion) [inline nilearn/_utils/data_gen.py:756]: if the function is meant to be part of the public API, it should have a docstring.
Otherwise the function name should have a leading underscore

File: nilearn/_utils/data_gen.py
Comment (bthirion) [inline nilearn/_utils/data_gen.py:760]: On which grounding are this subject and session id chosen ?

Comment (bthirion) [reply nilearn/_utils/data_gen.py:760]: I guess it's because this is only used for testing. Maybe indicate this in a short docstring.

File: nilearn/_utils/tests/test_data_gen.py
Comment (bthirion) [inline nilearn/_utils/tests/test_data_gen.py:25]: I'm suprised by the code indentation here. What magic leads to that ?

File: nilearn/_utils/tests/test_data_gen.py
Comment (bthirion) [inline nilearn/_utils/tests/test_data_gen.py:40]: Same thing here.

File: nilearn/glm/first_level/first_level.py
Comment (bthirion) [inline nilearn/glm/first_level/first_level.py:897]: Actually, it's not a percentage  it's a fraction

File: nilearn/glm/tests/test_first_level.py
Comment (bthirion) [inline nilearn/glm/tests/test_first_level.py:583]: The naming is great, but i'd not be against a short docstring...

File: nilearn/glm/tests/test_first_level.py
Comment (bthirion) [inline nilearn/glm/tests/test_first_level.py:765]: Sorry, but where does this 1.5 come from ?

File: nilearn/interfaces/tests/test_bids.py
Comment (bthirion) [inline nilearn/interfaces/tests/test_bids.py:30]: a one-linear to explain what this does ?

File: nilearn/_utils/data_gen.py
Comment (Remi-Gau) [reply nilearn/_utils/data_gen.py:760]: good point

File: nilearn/_utils/tests/test_data_gen.py
Comment (Remi-Gau) [reply nilearn/_utils/tests/test_data_gen.py:25]: Until #3614 is in, the files under the _utils folder will not checked by flake8 so "technically valid albeit interesting" formats are to be expected.

We could wait for #3614 to be in before we merge this PR, or the other way around: I will fix those formatting you are flagging here, but just want to let you know that any "format drift" you see should be fixed soon.

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [reply nilearn/glm/first_level/first_level.py:897]: indeed

File: nilearn/glm/tests/test_first_level.py
Comment (Remi-Gau) [reply nilearn/glm/tests/test_first_level.py:765]: ah yeah that definitely feels like a magic number: this is because the function that generate a BIDS dataset used for testing creates timeseries with a TR of 1.5 seconds: will explain than in the test docstring

File: nilearn/_utils/data_gen.py
Comment (Remi-Gau) [reply nilearn/_utils/data_gen.py:756]: done

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [reply nilearn/glm/first_level/first_level.py:897]: changed it also in the FirstLevelModel doc string

File: nilearn/glm/tests/test_first_level.py
Comment (Remi-Gau) [reply nilearn/glm/tests/test_first_level.py:581]: added some explanations: let me know if that's enough

File: nilearn/glm/tests/test_first_level.py
Comment (Remi-Gau) [reply nilearn/glm/tests/test_first_level.py:772]: added a comment

File: nilearn/interfaces/tests/test_bids.py
Comment (Remi-Gau) [reply nilearn/interfaces/tests/test_bids.py:35]: added some explanations: let me know if that's enough

File: nilearn/glm/first_level/first_level.py
Comment (bthirion) [inline nilearn/glm/first_level/first_level.py:816]: One-line docstring welcome

File: nilearn/glm/first_level/first_level.py
Comment (bthirion) [inline nilearn/glm/first_level/first_level.py:824]: One-line docstring welcome

File: nilearn/interfaces/tests/test_bids.py
Comment (bthirion) [inline nilearn/interfaces/tests/test_bids.py:92]: Can you add a docstring here ?

**Review by bthirion — COMMENTED** [review]
Almost there ! Thx

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [reply nilearn/glm/first_level/first_level.py:817]: done

File: nilearn/interfaces/tests/test_bids.py
Comment (Remi-Gau) [reply nilearn/interfaces/tests/test_bids.py:94]: done

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Comment by ymzayek** [comment]
@Remi-Gau can you rebase on main and resolve the conflict? Then this can be merged.

**Comment by Remi-Gau** [comment]
rebased + green CI = merging

**Pull request merged**
